python/cudf/cudf/core/series.py

```diff
@@ -815,6 +815,98 @@ def __array_function__(self, func, types, args, kwargs):
         else:
             return NotImplemented
 
+    def map(self, arg, na_action=None) -> "Series":
+        """
+        Map values of Series according to input correspondence.
+
+        Used for substituting each value in a Series with another value,
+        that may be derived from a function, a ``dict`` or
+        a :class:`Series`.
+
+        Parameters
+        ----------
+        arg : function, collections.abc.Mapping subclass or Series
+            Mapping correspondence.
+        na_action : {None, 'ignore'}, default None
+            If 'ignore', propagate NaN values, without passing them to the
+            mapping correspondence.
+
+        Returns
+        -------
+        Series
+            Same index as caller.
+
+        Examples
+        --------
+        >>> s = cudf.Series(['cat', 'dog', np.nan, 'rabbit'])
+        >>> s
+        0      cat
+        1      dog
+        2      NaN
+        3   rabbit
+        dtype: object
+
+        ``map`` accepts a ``dict`` or a ``Series``. Values that are not found
+        in the ``dict`` are converted to ``NaN``, unless the dict has a default
+        value (e.g. ``defaultdict``):
+
+        >>> s.map({'cat': 'kitten', 'dog': 'puppy'})
+        0   kitten
+        1    puppy
+        2      NaN
+        3      NaN
+        dtype: object
+
+        It also accepts a function:
+
+        >>> s.map('I am a {}'.format)
+        0       I am a cat
+        1       I am a dog
+        2       I am a nan
+        3    I am a rabbit
+        dtype: object
+
+        To avoid applying the function to missing values (and keep them as
+        ``NaN``) ``na_action='ignore'`` can be used:
+
+        >>> s.map('I am a {}'.format, na_action='ignore')
+        0     I am a cat
+        1     I am a dog
+        2            NaN
+        3  I am a rabbit
+        dtype: object
+        """
+        if isinstance(arg, dict):
+            result = self.copy().replace(arg)
+            for i in range(len(self)):
+                if self[i] not in arg.keys():
+                    result[i] = None
+        elif isinstance(arg, cudf.Series):
+            result = self.copy().replace(arg.index, arg)
+            for i in range(len(self)):
+                if self[i] not in arg.index:
+                    result[i] = None
+
+        elif isinstance(self._column, cudf.core.column.CategoricalColumn):
+            raise TypeError(
+                "User defined functions are currently not "
+                "supported on Series with dtypes `category`."
+            )
+        elif is_string_dtype(self._column.dtype):
+            if callable(arg):
+                result = self.copy()
+                for i in range(len(self)):
+                    result[i] = arg(self[i])
```

Comment:
This is going to do a device --> host copy for each element in the column which would make performance slow down to a crawl. We either need to apply the function at the column level (as opposed to the element level), or throw an Exception here.

Reply:
I see, I'll try out some options at the column level.

Reply:
I tried but wasn't able to figure out how to apply the function at the column level. I added an exception for both string and category callables.